### PR TITLE
Include a token on all Initial packets

### LINKED ISSUE #1649: Subsequent Initial Packets with Token Field
I was looking at the spec and it looks like we never explicitly indicated if subsequent Initial packets sent from the client should continue to populate the Token field. by subsequent packets, I mean responses to HRR and ACK only packets.

For the end server, I would assume it doesn't need the token for subsequent Initial packets, since it likely already has the state needed saved. But for a DDoS mitigation device in path, what is its logic supposed to be for these additional Initial packets it sees, assuming it wants to maintain a stateless design?

The DDoS mitigation device cannot really distinguish between the first Initial packet and subsequent ones. If the client always includes the token, the device would be able to statelessly verify it for every Initial packet, but that breaks if the end server also did a Retry. We could instead just assume full length (>1200 bytes) Initial packets are the first ones, and only they need to be validated. The device would just assume shorter packets should be let through untouched. I'm just not sure if that is too brittle an assumption.

### PR DESCRIPTION
**Pull request by martinthomson**
Closes #1649.

### CODE REVIEW

**Comment by mikkelfj** [comment]
What if the Intial packet with token needs to be sent again because of suspected packet loss?
(There is a lot of new text in this area, so it could well be covered).

**Comment by nibanks** [comment]
I agree with @mikkelfj here. If the client's first Initial (with a token) is suspected to be lost and the contents retransmitted, the token should continue to be included, right? Additionally, is there **harm** in including the token in all client Initial packets?

File: draft-ietf-quic-transport.md [hunk 1]
```diff
@@ -621,10 +621,19 @@ packet. It also sets the Token field to the token provided in the Retry. The
 client MUST NOT change the Source Connection ID because the server could include
 the connection ID as part of its token validation logic (see {{tokens}}).
 
+The Initial packet sent by the client in response to a Retry packet is subject
+to the same restrictions as the first Initial packet, with the exception of the
+value of the Destination Connection ID and Token fields, which are set as
+described here.  A client can either reuse the cryptographic handshake message
+or construct a new one at its discretion.  Any subsequent Initial packets from
+the client MUST use the same connection ID values, and MUST NOT include the
```
Comment (ianswett) [inline draft-ietf-quic-transport.md]: I thought the Retry now must change the CID?

**Comment by martinthomson** [comment]
Token on every Initial seemed wrong, but the more I think about it, the less reason I have for rejecting it.  Does anyone have a reason to remove the token?

File: draft-ietf-quic-transport.md (hunk 1 above)
Comment (martinthomson) [reply draft-ietf-quic-transport.md]: Yeah, I meant "same as the one from the Retry"  clarified.

File: draft-ietf-quic-transport.md [hunk 2]
```diff
@@ -621,10 +621,18 @@ packet. It also sets the Token field to the token provided in the Retry. The
 client MUST NOT change the Source Connection ID because the server could include
 the connection ID as part of its token validation logic (see {{tokens}}).
 
+The Initial packet sent by the client in response to a Retry packet is subject
+to the same restrictions as the first Initial packet, with the exception of the
+value of the Destination Connection ID and Token fields, which are set as
+described here.  A client can either reuse the cryptographic handshake message
```
Comment (ianswett) [inline draft-ietf-quic-transport.md]: How about putting the last sentence in this paragraph first, then saying "Aside from this, the Initial packet sent by the client is subject to the same restrictions as the first Initial packet."

**Comment by mikkelfj** [comment]
How does packet numbers relate to this?

How about requiring initial packets to be retransmitted verbatim and perhaps only allow token in the packet number 0? An endpoint can either store the packet for retransmission or reproduce the same AEAD on retransmission using the same packet number.

**Comment by ianswett** [comment]
If you retransmit verbatim, then you can't tell if the original or the retransmission was received, so you lose the first RTT sample on the client side, which is unfortunate, since that's the first congestion control or recovery related info you get on the connection.

**Pull request merged**